### PR TITLE
GEODE-9854: Orphaned .drf file causing memory leak

### PR DESCRIPTION
**Pull request by jvarenina**
Issue:
An OpLog files are compacted, but the .drf file is left because it contains deletes of
entries in previous .crfs. The .crf file is deleted, but the orphaned .drf is not until all
previous .crf files (.crfs with smaller id) are deleted.

The problem is that compacted Oplog object representing orphaned .drf file holds
a structure in memory (Oplog.regionMap) that contains information that is not useful
after the compaction and it takes certain amount of memory. Besides, there is a special
case when creating .krf files that, depending on the execution order,
could make the problem more severe (it could leave pendingKrfTags structure
on the regionMap and this could take up a significant amount of memory). This
pendingKrfTags HashMap is actually empty, but consumes memory because it was used
previously and the size of the HashMap was not reduced after it is cleared.
This special case usually happens when new Oplog is rolled out and previous Oplog
is immediately marked as eligible for compaction. Compaction and .krf creation start at
the similar time and compactor cancels creation of .krf file.
The pendingKrfTags structure is usually cleared when .krf file is created, but since
compaction canceled creation of .krf, the pendingKrfTags structure remain in memory
until Oplog representing orphaned .drf file is deleted.

Solution:
Clear the regionMap data structure of the Oplog when it is compacted (currently it is
deleted when the Oplog is destroyed).

Co-authored-by: Alberto Gomez <alberto.gomez@est.tech>

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java:74]: These are extraneous.

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java:94]: This is redundant -- cache.close() already calls it.

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java:119]: Either cast to `InternalRegion` or change `createRegion` to return `InternalRegion`.

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java:204]: `InternalRegion`

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java:455]: Maybe use local `DiskStoreFactoryImpl` variable or change incoming parameter to `DiskStoreFactoryImpl` in the method signature.

**Review by kirklund — CHANGES_REQUESTED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java:5722]: This line is the only difference than the previous pull request. You moved to handleEmpty() from cleanupAfterCompaction()?

File: geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java
Comment (jvarenina) [reply geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java:7666]: That was one of the proposed solution, but it was changed to this current one after @dschneider-pivotal change request https://github.com/apache/geode/pull/7145#discussion_r761565133

File: geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java
Comment (jvarenina) [reply geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java:5722]: Yes. this is the only difference in solution. We have moved regionMap close in handleEmpty() to cover one edge scenario that was not covered with the reverted PR. Also we have added new test case for the scenario. Please find more information's about it in comment section in ticket https://issues.apache.org/jira/browse/GEODE-9903.

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java
Comment (jvarenina) [reply geode-core/src/integrationTest/java/org/apache/geode/internal/cache/DiskRegionCompactorClearsObjectThatAreNoLongerNeededIntegrationTest.java]: Thank you for the review! I have applied all of your comments in latest commit.

File: geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java
Comment (DonalEvans) [reply geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java:7666]: Thanks for the clarification.

**Review by DonalEvans — APPROVED** [review]

**Comment by jvarenina** [comment]
Hi @kirklund , sorry for bothering. If you have time now, could you please check the latest commit where I have applied all of your requested changes.

**Comment by jvarenina** [comment]
Hi @kirklund , could you please review updates that are done after your changes requested. Changes requested is blocking this PR for a really long time now even all requested changes were applied.

**Review by mkevo — APPROVED** [review]

**Review by mivanac — APPROVED** [review]

**Comment by mivanac** [comment]
Hi, I am not sure what should be procedure when person who puts "Changes requested" is not reviewing requested changes.
What should be time limit, for waiting.

**Review by nabarunnag — APPROVED** [review]

**Review by kirklund — APPROVED** [review]
Approved the changes I requested. I still think someone with deeper oplog knowledge should review this PR.

**Review by gesterzhou — APPROVED** [review]

**Pull request merged**
